handlers/secrets.go

```diff
@@ -28,30 +29,75 @@ const (
 
 // MakeSecretHandler makes a handler for Create/List/Delete/Update of
 //secrets in the Kubernetes API
-func MakeSecretHandler(namespace string, kube kubernetes.Interface) http.HandlerFunc {
+func MakeSecretHandler(defaultNamespace string, kube kubernetes.Interface) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		if r.Body != nil {
 			defer r.Body.Close()
 		}
 
-		client := kube.CoreV1().Secrets(namespace)
+		lookupNamespace, err := getLookupNamespace(defaultNamespace, r)
+		if err != nil {
+			switch err.Error() {
+			case "unable to unmarshal Secret request":
+				http.Error(w, err.Error(), http.StatusBadRequest)
+			case "unable to manage secrets within the kube-system namespace":
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+			default:
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+		}
+
+		client := kube.CoreV1().Secrets(lookupNamespace)
 
 		switch r.Method {
 		case http.MethodGet:
 			getSecretsHandler(client, w, r)
 		case http.MethodPost:
-			createSecretHandler(client, namespace, w, r)
+			createSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodPut:
-			replaceSecretHandler(client, namespace, w, r)
+			replaceSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodDelete:
-			deleteSecretHandler(client, namespace, w, r)
+			deleteSecretHandler(client, lookupNamespace, w, r)
 		default:
 			w.WriteHeader(http.StatusBadRequest)
 			return
 		}
 	}
 }
 
+// getLookupNamespace returns the namespace specified in any request type
+func getLookupNamespace(defaultNamespace string, r *http.Request) (string, error) {
+	req := struct{ Namespace string }{Namespace: defaultNamespace}
```

Comment:
I am finding this confusing. I know Lucas suggested it, but why do we need this over a named type?

Reply:
This allows you to unmarshal any request that contains a Namespace field, and not  just one particular request type.

Reply:
I think mainly because we didn't have the `Secret` type from `faas-provider` with the new `Namespace` field. Now we do. Will update it.